### DISCUSSION TITLE
Notifications Rewrite

### DISCUSSION THREAD

**Discussion by mag37** [mag37/dockcheck#101] [General]
I've rewritten the new `releasenotes` function in the main script, to add to a new array straight away instead of just printf'ing the items. This due to some issues with different formatting of newlines and cleaner way to handle it with or without releasenotes-addition.

Hence rewritten this section in the templates (same in all templates):
```bash
[ -s "$ScriptWorkDir"/urls.list ] && releasenotes || Updates=("$@")
UpdToString=$( printf '%s\\n' "${Updates[@]}" )
```

Then I also rewrote most templates **MessageBody** - bits, this is what I fear might break some templates.

There's a branch here: [notify-rewrite](https://github.com/mag37/dockcheck/tree/notify-rewrite).

I've changed all templates to what I think will work, but I need to go through and test them all.

---

Those who run any notification addons are more then welcome to help with testing.

You need both the new `dockcheck.sh` and the `notify_xyz.sh` template of your choice.

I suggest you do it in a different directory and not overwrite your current setup.
If you're using git you can clone the branch: `git clone -b notify-rewrite https://github.com/mag37/dockcheck.git` in a new directory.
____

### List of current notification templates (will update when tested):
:white_check_mark: notify_apprise.sh
:white_check_mark: notify_discord.sh
:white_check_mark: notify_DSM.sh
:white_check_mark: notify_generic.sh
:white_check_mark: notify_gotify.sh
:white_check_mark: notify_matrix.sh
:white_check_mark: notify_ntfy-sh.sh
:white_check_mark: notify_pushbullet.sh
:white_check_mark: notify_pushover.sh
:white_check_mark: notify_smtp.sh
:white_check_mark: notify_telegram.sh

**Comment by mag37** [comment]
In most templates I've got it down to these:    

alt1 with \n still "intact" (works good for eg. json curls)
`MessageBody="🐋 Containers on $FromHost with updates available: \n$UpdToString"`

alt2 - where newlines needs to be processed already :
`MessageBody=$(printf "🐋 Containers on $FromHost with updates available: \n$UpdToString")`

**Comment by mag37** [comment]
@yoyoma2 
Maybe you could check out DSM/smtp? With the EOF - I'm not entirely sure how it'll handle the current variable.  

If it doesnt work as is maybe it has to be rewritten to set a variable before the EOF like:   
`MessageBody=$(printf "🐋 Containers on $FromHost with updates available: \n$UpdToString")`   
and use that **$MessageBody** instead of the current **$UpdToString**

**Comment by mag37** [comment]
Other people that might be able to test:

- @silverbios - matrix
- @cerealconyogurt - gotify
- @sudo-kraken - pushover
- @arpanghosh8453 - pushbullet

**Comment by yoyoma2** [comment]
Tried it on DSM and to get the identical email as before I had to delete an extra space in the script and delete a backslash from the DSM template. Did not try the smtp template but it has the same double backslash as the DSM template.
```
diff --git a/dockcheck.sh b/dockcheck.sh
index 83a5998..dda9efc 100755
--- a/dockcheck.sh
+++ b/dockcheck.sh
@@ -152,7 +152,7 @@ releasenotes() {
   for update in ${GotUpdates[@]}; do
     found=false
     while read -r container url; do
-      [[ $update == $container ]] && Updates+=("$update  ->   $url") && found=true
+      [[ $update == $container ]] && Updates+=("$update  ->  $url") && found=true
     done < "$ScriptWorkDir"/urls.list
     [[ $found == false ]] && Updates+=("$update  ->  url missing") || continue
   done
diff --git a/notify_templates/notify_DSM.sh b/notify_templates/notify_DSM.sh
index 84ba7e9..38a6ea7 100644
--- a/notify_templates/notify_DSM.sh
+++ b/notify_templates/notify_DSM.sh
@@ -19,7 +19,7 @@ fi

 send_notification() {
 [ -s "$ScriptWorkDir"/urls.list ] && releasenotes || Updates=("$@")
-UpdToString=$( printf '%s\\n' "${Updates[@]}" )
+UpdToString=$( printf '%s\n' "${Updates[@]}" )
 FromHost=$(hostname)
 CfgFile="/usr/syno/etc/synosmtp.conf"
```

**Reply by mag37** [reply]
Wonderful thank you for testing!
Was the `Updates+=("$update  ->   $url")` space causing troubles, or just "cosmetics"? Didnt realise it was 3 spaces there - I will change it back to 2. Good call.

yes the `'%s\\n'` double backslash within single-quotes is to escape it properly through the different interpretations. 
If you didn't remove the backslash did you get `\n` in the output instead of line breaks?

Would you mind trying this modification (with the  `'%s\\n'` in the assignment bit):

```
MessageBody=$(printf "🐋 Containers on $FromHost with updates available: \n$UpdToString")

$MailPkg $SendMailTo << __EOF
From: "$SenderName" <$SenderMail>
date:$(date -R)
To: <$SendMailTo>
Subject: $SubjectTag Updates available on $FromHost
Content-Type: text/plain; charset=UTF-8; format=flowed
Content-Transfer-Encoding: 7bit

$MessageBody

 From $SenderName
__EOF
}
```

**Reply by yoyoma2** [reply]
I was clicking back and forth between the two emails and any movement was visible. Two spaces before and after the arrow, just cosmetic, I like symmetry. :grin:

The \n was appearing as text (2 characters) on Thunderbird for Windows and the color of the link (including the \n) was for never visited and gave a 404 error when clicked. I didn't try any other email client or even Thunderbird on linux. DSM is basically a tweaked linux.

**Reply by mag37** [reply]
> Two spaces before and after the arrow, just cosmetic, I like symmetry. :grin:

Me too :) I'll fix it.

If you've got time please report back how it works with the example I added above.

**Reply by yoyoma2** [reply]
The newline after the "updates available:" wasn't working and the first container was ending up on the same line. :confused: By printing directly to the variable the result is more predictable. Had to delete an empty line following the use of the $MessageBody variable as it already contains a newline at the end.
```
printf -v MessageBody "🐋 Containers on $FromHost with updates available:\n\n$UpdToString"

$MailPkg $SendMailTo << __EOF
From: "$SenderName" <$SenderMail>
date:$(date -R)
To: <$SendMailTo>
Subject: $SubjectTag Updates available on $FromHost
Content-Type: text/plain; charset=UTF-8; format=flowed
Content-Transfer-Encoding: 7bit

$MessageBody
 From $SenderName
__EOF
}
```
Cool blue whale BTW. Is that going to work on all platforms?

**Reply by mag37** [reply]
Ah thank you for testing that out. And using `printf -v` is much cleaner than using the subshell to set the variable! I've just tested all other templates (that I have access to) and that works fine, so I changed them too.

Odd that the 2nd `\n` was needed for the linebreak after "updates available" but good find. I'll add that change to the templates and consider them working then! Thank you for testing, it's worth a lot - I'm just doing this work when I find time.

**Reply by yoyoma2** [reply]
I was clicking back and forth between the two emails and the second \n was to get things identical (the empty line before the first container). With `printf -v` the first \n did successfully end the line right after "updates available:".

I'm getting useful free software so testing a little is the least I can do.

**Reply by mag37** [reply]
> I was clicking back and forth between the two emails and the second \n was to get things identical (the empty line before the first container). With `printf -v` the first \n did successfully end the line right after "updates available:".
> 
> I'm getting useful free software so testing a little is the least I can do.

Ah so that was also cosmetic with an extra newline after the "updates available"-bit. Good to know!

Well thank you again for the feedback, really appreciated. Gonna consider some more changes then I'll merge to main asap!

**Comment by mag37** [comment]
I've now done the edits from feedback and testing to the branch. 

Still have these to test:
- matrix (@silverbios)
- gotify (@cerealconyogurt)
- pushbullet (@arpanghosh8453)

I'll hope for some of the contributors to comment, otherwise I'll try to setup testing myself or push as is and modify further on if needed.

**Reply by silverbios** [reply]
From the matrix(synapse), everything works fine with your last change without issues.

this an example of a message in synapse

🐋 Containers on $HOSTNAME with updates available: 
Container01
Container02

Thank you

**Reply by mag37** [reply]
Wonderful thank you for testing!

**Comment by mag37** [comment]
I spun up a gotify docker instance. Worked!

Logged in and set up a pushbullet token and tested. Worked!

So all templates are now confirmed working as advertised. I'll keep the discussion open if there's any feedback, suggestions or issues (though maybe open a issue if it's breaking).

Thank you guys!
